### PR TITLE
2 process queue

### PR DESCRIPTION
**Pull request by mdayican**
JIRA link (if applicable)
https://tools.hmcts.net/jira/browse/BPS-765

Change description

get message from queue and finalise it, dead letter it

Does this PR introduce a breaking change? (check one with "x")

[ ] Yes
[x] No

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    private void deadLetterTheMessage(
+        IMessage message,
+        String reason,
+        String description
+    ) throws InterruptedException, ServiceBusException {
+        messageReceiver.deadLetter(
+            message.getLockToken(),
+            reason,
+            description,
+            ImmutableMap.of("deadLetteredAt", Instant.now().toString())
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: do we need this deadLetteredAt ? that is present in orchestrator for a specific reason as we use to delete the message based on this field.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package uk.gov.hmcts.reform.bulkscan.payment.processor.service.servicebus.handler;
+
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.bulkscan.payment.processor.client.payhub.PayHubClientException;
+import uk.gov.hmcts.reform.bulkscan.payment.processor.service.servicebus.model.PaymentMessage;
+
+
+@Service
+public class PaymentMessageHandler {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java]: Either this class can be introduced by @aliveni as she is working on it or we can throw NotImplemeted exception in the below method.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/model/PaymentInfo.java [hunk 3]
```diff
@@ -12,5 +12,8 @@ public PaymentInfo(@JsonProperty(value = "document_control_number", required = t
         this.documentControlNumber = documentControlNumber;
     }
 
-
+    @Override
+    public String toString() {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/model/PaymentInfo.java]: do we need this ?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                break;
+            case UNRECOVERABLE_FAILURE:
+                deadLetterTheMessage(
+                    message,
+                    "Message processing error",
+                    processingResult.exception.getMessage()
+                );
+                break;
+            case POTENTIALLY_RECOVERABLE_FAILURE:
+                // starts from 0
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: nicer if we can extract code for this case clause into a separate method

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java (hunk 2 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/handler/PaymentMessageHandler.java]: added

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/model/PaymentInfo.java (hunk 3 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/model/PaymentInfo.java]: to ease logging.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        } catch (InvalidMessageException ex) {
+            log.error("Rejected message with ID {}, because it's invalid", message.getMessageId(), ex);
+            return new MessageProcessingResult(UNRECOVERABLE_FAILURE, ex);
+        } catch (PayHubClientException ex) {
+            logMessageProcessingError(message, payment, ex);
+
+            MessageProcessingResultType messageProcessingResultType = POTENTIALLY_RECOVERABLE_FAILURE;
+            if (ex.getStatus() == HttpStatus.BAD_REQUEST.CONFLICT) {
+                messageProcessingResultType = SUCCESS;
+            }
```
Comment (apereverzin) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: MessageProcessingResultType messageProcessingResultType = ex.getStatus() == HttpStatus.BAD_REQUEST.CONFLICT ? SUCCESS : POTENTIALLY_RECOVERABLE_FAILURE

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java (hunk 1 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: removed

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java (hunk 4 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: moved

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java (hunk 5 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: changed

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                break;
+            case UNRECOVERABLE_FAILURE:
+                deadLetterTheMessage(
+                    message,
+                    "Message processing error",
+                    processingResult.exception.getMessage()
+                );
+                break;
+            case POTENTIALLY_RECOVERABLE_FAILURE:
+                deadLetterIfDeliveryOver(message);
```
[the remark this answers is not in the document]
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: changed

File: build.gradle [hunk 7]
```diff
@@ -176,7 +176,8 @@ def versions = [
   junitPlatform   : '1.5.2',
   reformLogging   : '5.1.0',
   springBoot      : springBoot.class.package.implementationVersion,
-  springfoxSwagger: '2.9.2'
+  springfoxSwagger: '2.9.2',
+  mockitoJupiter  : '3.0.0'
```
Comment (doncem) [inline build.gradle]: only used once - no need to stick in defs

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    private MessageProcessingResult process(IMessage message) {
+        log.info("Started processing message with ID {}", message.getMessageId());
+
+        PaymentMessage payment = null;
+
+        try {
+            payment = paymentMessageParser.parse(message.getMessageBody());
+            logMessageParsed(message, payment);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: we could log message in parser itself as a completion step

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+            paymentMessageHandler.handlePaymentMessage(payment);
+            log.info("Processed message with ID {}. Envelope ID: {}", message.getMessageId(), payment.envelopeId);
+            return new MessageProcessingResult(SUCCESS);
+        } catch (InvalidMessageException ex) {
+            log.error("Rejected message with ID {}, because it's invalid", message.getMessageId(), ex);
+            return new MessageProcessingResult(UNRECOVERABLE_FAILURE, ex);
+        } catch (PayHubClientException ex) {
+            logMessageProcessingError(message, payment, ex);
+            MessageProcessingResultType messageProcessingResultType =
+                (ex.getStatus() == HttpStatus.BAD_REQUEST.CONFLICT) ? SUCCESS : POTENTIALLY_RECOVERABLE_FAILURE;
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: huh? `HttpStatus.CONFLICT` should be enough shouldn't it? And then `BAD_REQUEST` - unrecoverable. the rest - recoverable¿

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: \+ if it is success, perhaps `logMessageProcessingError` is not quite right? as it is success `log.error` wouldn't really be relevant

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        String reason,
+        String description
+    ) throws InterruptedException, ServiceBusException {
+        messageReceiver.deadLetter(
+            message.getLockToken(),
+            reason,
+            description
+        );
+
+        log.info("Message with ID {} has been dead-lettered", message.getMessageId());
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: reason could be handy

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    private void logMessageFinaliseError(
+        IMessage message,
+        MessageProcessingResultType processingResultType,
+        Exception ex) {
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: `)` should be in the new line ;)

File: src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java (hunk 10 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java]: done

**Pull request merged**